File: python/01-arrays-hashing/max_char.py
"""
Problem: Max Char
Pattern: Arrays & Hashing (frequency counting)
Time: O(n) | Space: O(n)

Given a string, return the character used most often in it.

Input:  max_char('abcccccccd')
Output: 'c'
"""
import re
from collections import Counter

def max_char(input_str):
    cleaned = re.sub(r'[^\w]', '', input_str)
    counts = Counter(cleaned)
    return max(counts, key=counts.get) if counts else ''


# Alternative: sort characters, then count runs of equal adjacent chars in
# one pass. Time: O(n log n) instead of O(n), but avoids building an
# explicit frequency map -- worth mentioning as a trade-off.
def max_char_sorted(input_str):
    chars = sorted(re.sub(r'[^\w]', '', input_str))
    max_char_val, max_count = '', 0
    current_char, current_count = '', 0
    for c in chars:
        if c == current_char:
            current_count += 1
        else:
            current_char, current_count = c, 1
        if current_count > max_count:
            max_count, max_char_val = current_count, current_char
    return max_char_val


if __name__ == "__main__":
    print(max_char('abcccccccd'))  # c

File: javascript/01-arrays-hashing/maxChar.js
// --- Directions
// Given a string, return the character that is most
// commonly used in the string.
// --- Examples
//
// Approach 1: hash map of counts, then linear scan for the max (below)
// Approach 2 (alternative): sort characters, then count runs of equal
// adjacent chars in one pass -- O(n log n) instead of O(n), but avoids
// building an explicit frequency map. Useful to mention as a trade-off
// in interviews even though approach 1 is strictly better here.

function maxChar(input) {
    const charCount = {};
    input.replace(/[^\w]/g, '').split('').forEach(char => {
        charCount[char] = (charCount[char] || 0) + 1;
    });
  
    let maxChar = '';
    let maxCount = 0;
  
    for (const char in charCount) {
        if (charCount[char] > maxCount) {
            maxCount = charCount[char];
            maxChar = char;
        }
    }

    //`{ ${maxChar} : ${maxCount} }`
  
    return maxChar; 
}

// Alternative: sort + count runs. Time: O(n log n) | Space: O(n)
function maxCharSorted(input) {
    const chars = input.replace(/[^\w]/g, '').split('').sort();
    let maxChar = '';
    let maxCount = 0;
    let currentChar = '';
    let currentCount = 0;

    for (const char of chars) {
        if (char === currentChar) {
            currentCount++;
        } else {
            currentChar = char;
            currentCount = 1;
        }
        if (currentCount > maxCount) {
            maxCount = currentCount;
            maxChar = currentChar;
        }
    }
    return maxChar;
}

module.exports = maxChar;
module.exports.alternative = maxCharSorted;
  
